taskchampion/taskchampion/src/server/cloud/server.rs

```
@@ -0,0 +1,1130 @@
+    cryptor: Cryptor,
+
+    /// The probability (0..255) that this run will perform cleanup.
+    cleanup_probability: u8,
+
+    /// For testing, a function that is called in the middle of `add_version` to simulate
+    /// a concurrent change in the service.
+    #[cfg(test)]
+    add_version_intercept: Option<fn(service: &mut SVC)>,
+}
+
+const LATEST: &[u8] = b"latest";
+const DEFAULT_CLEANUP_PROBABILITY: u8 = 13; // about 5%
+
+#[cfg(not(test))]
+const MAX_VERSION_AGE_SECS: u64 = 3600 * 24 * 180; // about half a year
+
+fn version_to_bytes(v: VersionId) -> Vec<u8> {
+    v.as_simple().to_string().into_bytes()
+}
+
+impl<SVC: Service> CloudServer<SVC> {
+    pub(in crate::server) fn new(service: SVC, encryption_secret: Vec<u8>) -> Result<Self> {
+        let cryptor = Cryptor::new(b"Taskchampion", &encryption_secret.into())?;
+        Ok(Self {
+            service,
+            cryptor,
+            cleanup_probability: DEFAULT_CLEANUP_PROBABILITY,
+            #[cfg(test)]
+            add_version_intercept: None,
+        })
+    }
+
+    /// Generate an object name for the given parent and child versions.
+    fn version_name(parent_version_id: &VersionId, child_version_id: &VersionId) -> Vec<u8> {
+        format!(
+            "v-{}-{}",
+            parent_version_id.as_simple(),
+            child_version_id.as_simple()
+        )
+        .into_bytes()
+    }
+
+    /// Parse a version name as generated by `version_name`.
+    fn parse_version_name(name: &[u8]) -> Option<(VersionId, VersionId)> {
+        if name.len() != 2 + 32 + 1 + 32 || !name.starts_with(b"v-") || name[2 + 32] != b'-' {
+            return None;
+        }
+        let Ok(parent_version_id) = VersionId::try_parse_ascii(&name[2..2 + 32]) else {
+            return None;
+        };
+        let Ok(child_version_id) = VersionId::try_parse_ascii(&name[2 + 32 + 1..]) else {
+            return None;
+        };
+        Some((parent_version_id, child_version_id))
+    }
```

Would it make sense to log a warning before each place we `return None` here? In practice it looks like only valid version names should be fed into this function and I'm not sure parsing should fail silently.

---

This might happen if the user adds other files to the bucket (which they shouldn't!). If that generates a warning, users will see a lot of them! Also, we don't have a way to log user-visible warnings here.  `log::warn!` won't appear unless the log level is set properly, which is currently impossible (#3220). I'll add a comment as to when None is returned here.